lib/pages/add_page/add_page.dart

```diff
@@ -40,11 +45,80 @@ class _AddPageState extends ConsumerState<AddPage> with Functions {
     super.dispose();
   }
 
+  String getCleanAmountString() {
+    // Remove all non-numeric characters
+    var cleanNumberString =
+        amountController.text.replaceAll(RegExp(r'[^0-9\.]'), '');
+
+    // Remove leading zeros
+    return cleanNumberString.replaceAll(RegExp(r'^[0\.]+(?=.)'), '');
+  }
+
+  void _updateAmount() {
+    final selectedType = ref.read(transactionTypeProvider);
+
+    var toBeWritten = getCleanAmountString();
+
+    if (selectedType == TransactionType.expense) {
+      // apply the minus sign if it's an expense
+      if (toBeWritten.isNotEmpty) {
+        toBeWritten = "-$toBeWritten";
+      }
+    }
+
+    if (toBeWritten != amountController.text) {
+      // only update the controller if the value is different
+      amountController.text = toBeWritten;
+      amountController.selection = TextSelection.fromPosition(
+        TextPosition(offset: toBeWritten.length),
+      );
+    }
+  }
+
+  void _createOrUpdateTransaction() {
+    final selectedType = ref.read(transactionTypeProvider);
+    final selectedTransaction = ref.read(selectedTransactionUpdateProvider);
+
+    final cleanAmount = getCleanAmountString();
+
+    // Check that an amount it's inserted
+    if (cleanAmount != '') {
+      if (selectedTransaction != null) {
+        ref
+            .read(transactionsProvider.notifier)
+            .updateTransaction(currencyToNum(cleanAmount), noteController.text)
+            .whenComplete(() => Navigator.of(context).pop());
+      } else {
+        if (selectedType == TransactionType.transfer) {
+          if (ref.read(bankAccountTransferProvider) != null) {
+            ref
+                .read(transactionsProvider.notifier)
+                .addTransaction(currencyToNum(cleanAmount), noteController.text)
+                .whenComplete(() => Navigator.of(context).pop());
+          }
+        } else {
+          // It's an income or an expense
+          if (ref.read(categoryProvider) != null) {
+            ref
+                .read(transactionsProvider.notifier)
+                .addTransaction(currencyToNum(cleanAmount), noteController.text)
+                .whenComplete(() => Navigator.of(context).pop());
+          }
+        }
+      }
+    }
+  }
+
   @override
   Widget build(BuildContext context) {
     final selectedType = ref.watch(transactionTypeProvider);
     final selectedTransaction = ref.watch(selectedTransactionUpdateProvider);
 
+    ref.listen(transactionTypeProvider, (previous, next) {
+      // update the amount text when the transaction type changes
+      _updateAmount();
+    });
+
```

Comment:
If you do it this way when you click on an expense transaction and the "Edit transaction" page opens up you are still missing the minus sign. To add it you have to first change the type of the transaction and then the `_updateAmount` function is called. 

So to be clear on the issue:

1. Click on an expense Transaction
2. The "Editing transaction" page opens up 
3. The value of the transaction is in red but is still missing the minus 

I believe that you could just don't use this callback and simply call the function: 

```suggestion
_updateAmount();
```

But maybe I'm missing something, let me know.

Reply:
It worked!